### PR TITLE
Add todaymanga

### LINKED ISSUE #1183: Source Request: Today Manga
### Source name

Today Manga 

### Source link

https://todaymanga.com/

### Source language

English

### Other details

_No response_

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [ ] I have admitted that I am a clown by having checked this box, as I have not read these acknowledgements.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by Secozzi**
Closes #1183

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+    override fun latestUpdatesNextPageSelector(): String =
+        popularMangaNextPageSelector()
+
+    // =============================== Search ===============================
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
+        val filterList = if (filters.isEmpty()) getFilterList() else filters
+        val categoryFilter = filterList.find { it is CategoryFilter } as CategoryFilter
+        val genreFilter = filterList.find { it is GenreFilter } as GenreFilter
```
Comment (AwkwardPeak7) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        val categoryFilter = filterList.filterIsInstance<CategoryFilter>().first()
        val genreFilter = filterList.filterIsInstance<GenreFilter>().first()
```

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                    addPathSegments("category/most-popular")
+                }
+            }
+
+            if (page > 1) {
+                addQueryParameter("page", page.toString())
+            }
+        }
+
+        return GET(url.build(), headers)
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        }.build()

        return GET(url, headers)
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+    open class UriPartFilter(displayName: String, private val vals: Array<Pair<String, String>>) :
+        Filter.Select<String>(displayName, vals.map { it.first }.toTypedArray()) {
+        fun toUriPart() = vals[state].second
+    }
+
+    // =========================== Manga Details ============================
+
+    override fun mangaDetailsParse(document: Document): SManga = SManga.create().apply {
+        document.selectFirst(".serie")!!.run {
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        with(document.selectFirst(".serie")!!) {
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    // ============================== Popular ===============================
+
+    override fun popularMangaRequest(page: Int): Request =
+        GET("$baseUrl/category/most-popular".addPage(page), headers)
+
+    override fun popularMangaSelector(): String = "section div.serie"
+
+    override fun popularMangaFromElement(element: Element): SManga = SManga.create().apply {
+        setUrlWithoutDomain(element.selectFirst("a[href]")!!.attr("abs:href"))
+        thumbnail_url = element.selectFirst("img")?.imgAttr()
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        thumbnail_url = element.selectFirst("img")!!.imgAttr()
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        GET("$baseUrl/category/recent".addPage(page), headers)
+
+    override fun latestUpdatesSelector(): String = "ul.series > li"
+
+    override fun latestUpdatesFromElement(element: Element): SManga = SManga.create().apply {
+        element.selectFirst("a[title][href]")!!.run {
+            setUrlWithoutDomain(attr("abs:href"))
+            title = attr("title")
+        }
+        thumbnail_url = element.selectFirst("img")?.imgAttr()
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        thumbnail_url = element.selectFirst("img")!!.imgAttr()
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        Filter.Select<String>(displayName, vals.map { it.first }.toTypedArray()) {
+        fun toUriPart() = vals[state].second
+    }
+
+    // =========================== Manga Details ============================
+
+    override fun mangaDetailsParse(document: Document): SManga = SManga.create().apply {
+        document.selectFirst(".serie")!!.run {
+            title = selectFirst("h1")!!.text()
+            thumbnail_url = selectFirst("img")?.imgAttr()
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
            thumbnail_url = selectFirst("img")!!.imgAttr()
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+            genre = select(".serie-info-head .tags > .tag-item").joinToString { it.text() }
+            author = select(".authors a").joinToString { it.text() }
+            status = selectFirst("li:contains(status) span").parseStatus()
+        }
+
+        description = buildString {
+            document.selectFirst(".serie-summary")?.also { summary ->
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]: Is this expected to be nullable? If not, use `val summary = ...!!`.

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            add("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
+            set("Referer", baseUrl + manga.url)
+        }.build()
+        return GET("$baseUrl${manga.url}/chapter-list", chapterHeaders)
+    }
+
+    override fun chapterListSelector() = "ul.chapters-list > li"
+
+    override fun chapterFromElement(element: Element): SChapter = SChapter.create().apply {
+        element.selectFirst("a")!!.run {
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        with(element.selectFirst("a")!!) {
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+        } else if (dateText.contains("ago")) {
+            dateText.parseRelativeDate()
+        } else {
+            parseDate(dateText)
+        }
+    }
+
+    private fun parseDate(dateStr: String): Long {
+        return runCatching { DATE_FORMATTER.parse(dateStr)?.time }
+            .getOrNull() ?: 0L
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        return try {
            DATE_FORMATTER.parse(dateStr)!!.time
        } catch (_: Throwable) {
            0L
        }
```

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+        hasAttr("data-lazy-src") -> attr("abs:data-lazy-src")
+        hasAttr("data-src") -> attr("abs:data-src")
+        else -> attr("abs:src")
+    }
+
+    companion object {
+        private val DATE_FORMATTER by lazy {
+            SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)
+        }
+    }
```
Comment (stevenyomi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
    private val DATE_FORMATTER by lazy {
        SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)
    }
```

**Review by stevenyomi — CHANGES_REQUESTED** [review]

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+            dateText.parseRelativeDate()
+        } else {
+            parseDate(dateText)
+        }
+    }
+
+    private fun parseDate(dateStr: String): Long {
+        return try {
+            dateFormat.parse(dateStr)!!.time
+        } catch (_: Throwable) {
```
Comment (beer-psi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        } catch (_: ParseException) {
```

[DateFormat#parse(String)](https://docs.oracle.com/javase/8/docs/api/java/text/DateFormat.html#parse-java.lang.String-) doesn't return null, so you only need to catch [ParseException](https://docs.oracle.com/javase/8/docs/api/java/text/ParseException.html). Catching too wide may suppress other issues.

File: src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+            // parses: "2 months ago"
+            "month" in this -> {
+                parsedDate = now.apply { add(Calendar.MONTH, -relativeDate) }.timeInMillis
+            }
+            // parse: "2 years ago"
+            "year" in this -> {
+                parsedDate = now.apply { add(Calendar.YEAR, -relativeDate) }.timeInMillis
+            }
+        }
+        return parsedDate
```
Comment (beer-psi) [inline src/en/todaymanga/src/eu/kanade/tachiyomi/extension/en/todaymanga/TodayManga.kt]:
```suggestion
        val relativeDate = this.split(" ").firstOrNull()
            ?.replace("one", "1")
            ?.replace("a", "1")
            ?.toIntOrNull()
            ?: return 0L

        when {
            // parse: 30 seconds ago
            "second" in this -> now.add(Calendar.SECOND, -relativeDate)

            // parses: "42 minutes ago"
            "minute" in this -> now.add(Calendar.MINUTE, -relativeDate)

            // parses: "1 hour ago" and "2 hours ago"
            "hour" in this -> now.add(Calendar.HOUR, -relativeDate)

            // parses: "2 days ago"
            "day" in this -> now.add(Calendar.DAY_OF_YEAR, -relativeDate)

            // parses: "2 weeks ago"
            "week" in this -> now.add(Calendar.WEEK_OF_YEAR, -relativeDate)

            // parses: "2 months ago"
            "month" in this -> now.add(Calendar.MONTH, -relativeDate)

            // parse: "2 years ago"
            "year" in this -> now.add(Calendar.YEAR, -relativeDate)
        }
        return now.timeInMillis
```

**Review by beer-psi — CHANGES_REQUESTED** [review]

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**